### DISCUSSION TITLE
Støtte read-only skjema-komponenter

### DISCUSSION THREAD

**Discussion by sjur-gr** [navikt/Designsystemet#1889] [Komponenter]
Vi ønsker å vurdere skjema-komponenter som er read-only. Har du innspill til hvordan vi kan støtte read-only på en god og brukervennlig måte?

**Comment by JorgenMaristuen** [comment]
## Hva er read-only?
Readonly-attributtet spesifiserer at et inputfelt i et skjema er skrivebeskyttet. Et skrivebeskyttet felt kan ikke endres. Til forskjell fra et disabled/deakivert felt, kan en bruker tabbe til det, markere det og kopiere teksten fra det. Verdiene i readonly-felter inkluderes også når et skjema sendes inn.

## Kartlegging av behov i NAV
Etter å ha snakket litt med teamene, ser vi at det finnes to forskjellige behov:

### 1. Enkel visning
Typisk brukt ved totrinnskontroll hos saksbehandlere. Lignende løsninger brukes også i oppsummering av søknadsløp på eksterne flater. Her er poenget å gjøre enkelt å lese/skanne innholdet uten unødvendig støy. Derfor fjernes vanligvis all grafisk styling og tomme datafelter, og det som vises er label og tekst/verdi. Avhengig av type input-felt vises verdiene som ren tekst, kommaseparerte verdier, eller kulepunkter. 
<details><summary>Eksempel på redigeringsmodus og lesemodus</summary>
<p>

Redigeringsmodus:
![image](https://github.com/navikt/aksel/assets/86107432/c20d3f67-e04a-4e7a-873a-7e529ad22378)

Lesemodus:
![image](https://github.com/navikt/aksel/assets/86107432/b78a50db-6147-4101-8d23-ed2ab98b8a56)
</p>
</details> 

På disse sidene/skjemaene er det ikke en blanding av redigerbare og read-only felter. Men i noen tilfeller kan det være mulighet for å åpne redigeringsmodus for et felt ved å klikke på et ikon e.l.:
<details><summary>Eksempel på lesemodus med redigeringsmulighet</summary>
<p>

Lesemodus med redigeringsmulighet:
![image](https://github.com/navikt/aksel/assets/86107432/82f837f4-b849-43fe-aa04-a0a950403470)

Redigeringsmodus åpnet:
![image](https://github.com/navikt/aksel/assets/86107432/104f8a22-db01-4a21-b7e5-71c9272a4235)

</p>
</details>

Noen ganger er det behov for å fjerne og omstrukturere skjemainnholdet for lesemodus, og da er spørsmålet om dette fortsatt er det samme skjemaet, eller en nye side med annen visning av dataene:

<details><summary>Eksempel på omstrukturert innhold</summary>
<p>

Redigeringsmodus:
![image](https://github.com/navikt/aksel/assets/86107432/42fa2d17-40cb-4dd5-b005-c0ff773a52d6)

Lesemodus:
![image](https://github.com/navikt/aksel/assets/86107432/71afcb9e-2609-4b18-acc0-74b8da475b4b)


</p>
</details> 


### 2. Nedtonet komponent-styling
I skjemaer som inneholder både felter som er redigerbare og felter som er skrivebeskyttet, kan det være et poeng å vise hva slags type input-felter som er skrivebeskyttet og hva slags valg som er mulige. Eksempelvis kan det være et skjema hvor det av tekniske årsaker ikke er mulig å redigere alle felter, men der dette vil bli mulig i fremtiden. 
<details><summary>Eksempel team Valp</summary>
<p>

Eksempel fra team Valp (ønsker konteksten ved at det fortsatt ser ut som input-felter med tanke på at vi på sikt skal fjerne read-only.)
![image](https://github.com/navikt/aksel/assets/86107432/9ebfde16-6fe3-472b-9150-1adb292bf85b)

</p>
</details> 

<details><summary>Eksempel Foreldrepenger</summary>
<p>

![image](https://github.com/navikt/aksel/assets/86107432/79db719a-fee8-4e64-95d0-cc2f38f56d41)

</p>
</details> 

## Hvordan kan Aksel gjøre det enklere?
Vi ser et ganske utbredt behov for "enkel visning" til oppsummeringer og totrinnskontroll. Fra noen har vi fått ønsker om egne props fra designsystemet. Andre styrer visningen i en wrapper for hver form-komponent som håndterer knytningen mellom react-hook-form og komponenten fra ds-react. Dersom det er behov for å fjerne eller omstrukturere mer av innholdet i en komponent, kan det bli teknisk vanskelig å lage dette som en prop. Et fleksibelt alternativ kan være en [summary list komponent](https://design-system.service.gov.uk/components/summary-list/) slik som gov.uk bruker.

Det andre behovet, "nedtonet komponent-styling", er det som det gir mest mening å ha som prop, siden vi her beholder det grunnleggende utseendet til hver komponent. Dette tenker vi da å lage som en egen state i Figma og kode, på linje med active, hover og disabled.

Skisse:
![image](https://github.com/navikt/aksel/assets/86107432/f12e6263-083e-412b-b8a8-d50eabfbd81a)

**Comment by gunnjohansen2** [comment]
Når det er klart, kunne dere lagt ut en lenke til skisser i Figma? 😊

**Reply by JorgenMaristuen** [reply]
Ja, det skal vi gjøre! 😃

**Comment by JorgenMaristuen** [comment]
Her kan dere teste ut komponenter med read-only: [Hjelp oss med Read-only i Figma](https://www.figma.com/file/htPczlWeORPi71eYsm1FFz/Read-only?type=design&node-id=201-26967&mode=design&t=nFfRF45fax16slna-4)
@gunnjohansen2
